### PR TITLE
MVC / View / Razor article ~ WIP

### PR DESCRIPTION
**Pull request by robbert229**
I am opening up this pr for now, but will rebase and make changes based on feedback. Feedback is appreciated. 

One area that I would like direction on is the ``@inherits`` directive. I feel kind of lost on to what I should be saying on it, and how deep I should dive into it.

A lot of this article is also not razor syntax specific but details on using features of MVC through razor. Is this topic the right place for this, or does it belong in another topic? 

@NTaylorMullen @Rick-Anderson @pranavkm @danroth27

### CODE REVIEW

File: mvc/views/razor.rst [hunk 1]
```diff
@@ -1,8 +1,283 @@
...
+Razor Syntax In MVC 6
+=====================
+By `John Rowley`_
 
-|stub-icon| Razor Syntax
-========================
+---------------
+What is Razor?
+---------------
+Razor is the ASP.NET view engine used in MVC. It allows c# and
```
Comment (pranavkm) [inline mvc/views/razor.rst]: Perhaps use the line from readme instead:

> The Razor syntax provides a fast, terse, clean and lightweight way to combine server code with HTML to create dynamic web content. This repo contains the parser and the C# code generator for the Razor syntax.

Razor isn't a view engine, we just happen to build a view engine that parses and compiles files using the Razor syntax.

File: mvc/views/razor.rst [hunk 2]
```diff
@@ -1,8 +1,283 @@
...
+
+.. code-block:: c#
+
+  <p>@(1 + 1) = 1 + 1</p>
+
+*********
+Encoding
+*********
+Rendered data is by default encoded so that any non-text characters will be
+escaped. To render unencodedhtml use ``@Html.Raw``.
```
Comment (pranavkm) [inline mvc/views/razor.rst]: unencoded HTML

File: mvc/views/razor.rst [hunk 3]
```diff
@@ -1,8 +1,283 @@
...
+.. code-block:: c#
+
+  ISBN: @Model.ISBN
+  Author: @Model.AuthorName()
+
+The ``@using`` directive is the Razor equivalent of the c# ``using`` statement.
+
+.. code-block:: c#
+
+  @using Example.Models;
```
Comment (pranavkm) [inline mvc/views/razor.rst]: `@using Example.Models` (no semicolon)

**Comment by danroth27** [comment]
@Eilon Please review

**Comment by NTaylorMullen** [comment]
> A lot of this article is also not razor syntax specific but details on using features of MVC through razor. Is this topic the right place for this, or does it belong in another topic?

After seeing the section that this article will lie in (Mvc => Views etc.) I think the MVC tid bits make sense.

File: mvc/views/razor.rst (hunk 1 above)
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: :+1:

File: mvc/views/razor.rst [hunk 4]
```diff
@@ -1,8 +1,283 @@
...
 
-.. include:: /../common/stub-notice.txt
+Razor is the only view engine that MVC ships, but it is possible to use a
+different view engine.
 
-.. _issue: https://github.com/aspnet/Docs/issues/130
+Implicit Expressions
+^^^^^^^^^^^^^^^^^^^^^
+Implicit Expressions provide a simple mechanism to render data. They are simply
+an ``@`` followed by a variable or method. Razor renders the value to at
```
Comment (pranavkm) [inline mvc/views/razor.rst]: > Razor renders the value to the output at

File: mvc/views/razor.rst [hunk 5]
```diff
@@ -1,8 +1,283 @@
...
+  @(Html.Raw(SuperSpecialHTML))
+
+``@Html.Raw`` is a Razor HTML Helper, for more information on them see
+:doc:`html-helpers`.
+
+*******
+Models
+*******
+The ``@model`` directive is a way to indicate what model the view is using. Just
+place it at the top of the view along with the class that you are wanting to
```
Comment (pranavkm) [inline mvc/views/razor.rst]: along with ~~class~~ type of the model.

File: mvc/views/razor.rst [hunk 6]
```diff
@@ -1,8 +1,283 @@
...
+    @{
+      var GuestName = "John";
+    }
+
+    <h2>Greetings @GuestName</h2>
+  </div>
+
+You can mix plain text, and html inside of a code block by using
+html tags and the ``<text>`` tag. The ``<text>`` tag is a placeholder tag that
+is only used server side to tell razor that the following content is text and
```
Comment (pranavkm) [inline mvc/views/razor.rst]: is a placeholder tag that is used to indicate that the enclosed content is text and should not be parsed as C# code.

**Comment by NTaylorMullen** [comment]
We had a few comments from the word doc Rick passed along earlier. I'll surface some of the comments here with names prefixed. Ex:

`Pranavkm: Hello World`

File: mvc/views/razor.rst [hunk 7]
```diff
@@ -1,8 +1,283 @@
...
+    <h1>@title</h1>
+
+    var body = "some text not sitting inside of a tag";
+    <text>@body</text>
+  }
+
+*******************
+Control Structures
+*******************
+A lot of the power behind razor comes from the ability to have control structures
```
Comment (pranavkm) [inline mvc/views/razor.rst]: I think `razor` should be cased as `Razor`. A couple of places in the file. So perhaps a find-and-replace?

**Comment by NTaylorMullen** [comment]
NTaylorMullen:
Should mention TagHelpers as a section and link to the corresponding docs article.

Should mention @section if you decide to go with MVC concepts in here.

Should mention @functions

pranavkm:
Might also want to add a note about how attribute conditionals workl. It’s implemented in Mvc but it’s essentially Razor behavior:
@{ var myclass = null; }
<span class=”@myclass”>Hello</span>

would result in the class attribute being dropped entirely.

File: mvc/views/razor.rst [hunk 8]
```diff
@@ -1,8 +1,283 @@
...
+Implicit Expressions
+^^^^^^^^^^^^^^^^^^^^^
+Implicit Expressions provide a simple mechanism to render data. They are simply
+an ``@`` followed by a variable or method. Razor renders the value to at
+runtime. The following markup renders the variable `` GuestName`` inside an H2
+heading element.
+
+.. code-block:: c#
+
+  <h2>@GuestName</h2>
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: NTaylorMullen:
Should probably show how the GuestName variable is set to showcase what it’s really doing here. I’d imagine a user reading this document would be completely unfamiliar with Razor and wouldn’t understand how `@GuestName` translated into “Guest-Name”.

File: mvc/views/razor.rst [hunk 9]
```diff
@@ -1,8 +1,283 @@
...
+
+  <a href="#" class="@AnchorClass" id="@AnchorId"> Link </a>
+
+The output for this markup is below.
+
+.. code-block:: html
+
+  <a href="#" class="dark link" id="main-link"> Link </a>
+
+When the value you pass into an attribute is null razor will remove that
```
Comment (pranavkm) [inline mvc/views/razor.rst]: is null or boolean false, Razor will remove

Comment (pranavkm) [inline mvc/views/razor.rst]: Nevermind. You specify that later on.

File: mvc/views/razor.rst [hunk 10]
```diff
@@ -1,8 +1,283 @@
...
+
+  <h2>@GuestName</h2>
+
+Razor automatically escapes the ``@`` used in an email address. To escape ``@``
+use ``@@``
+
+*********
+Explicit
+*********
+Sometimes you need to render variables without any space in between it and
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Your example later in this article `@await ____` conflicts with this statement. In essence `@await Foo` is a special implicit expression Syntax. May want to mention it as an exception to the rule.

File: mvc/views/razor.rst [hunk 11]
```diff
@@ -1,8 +1,283 @@
...
+surrounding text. This is where explicit expressions come in. With explicit
+expressions you surround the variable name with parenthesis to explicitly
+show that it is an expression instead of random text.
+
+.. code-block:: c#
+
+  <h2>@(MajorVersion)@(MinorVersion)</h2>
+
+Explicit expressions are evaluated like regular c#, as such it is possible
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: @Rick-Anderson What's the common casing for `c#` and `html` that we use in other documents? I saw an earlier lowercased `html` and `c#` is used throghout.

File: mvc/views/razor.rst (hunk 9 above)
Comment (pranavkm) [inline mvc/views/razor.rst]: There's a secondary aspect to this, it also helps in pretty-printing. For instance if you have

```
<a href="#" class="bootstrap-anchor @classes"> Link </a>
```

Razor will remove the space preceding the `@classes` value in the attribute (if it's null or false) so you'd get:
```
<a href="#" class="bootstrap-anchor"> Link </a>
```

File: mvc/views/razor.rst [hunk 12]
```diff
@@ -1,8 +1,283 @@
...
+to render math, and much more.
+
+.. code-block:: c#
+
+  <p>@(1 + 1) = 1 + 1</p>
+
+*********
+Encoding
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: You may want to nest Implicit and Explicit Expressions under a parent `Expressions` section. That way you can mention that all expressions are encoded by default and to opt-out of that behavior you can use `Html.Raw`. 

Having a separate encoding section like this doesn't entirely belong.

File: mvc/views/razor.rst [hunk 13]
```diff
@@ -1,8 +1,283 @@
...
+
+  @await FlushAsync()
+
+**********
+@inherits
+**********
+``@inherits`` Allows you to override the class that view inherits from. By
+default the view inherits from ``System.Web.Mvc.WebViewPage``.
+
+.. I am at a loss for words here...
```
Comment (pranavkm) [inline mvc/views/razor.rst]: Aww

File: mvc/views/razor.rst [hunk 14]
```diff
@@ -1,8 +1,283 @@
...
+  @Html.Raw(SuperSpecialHTML)
+  @(Html.Raw(SuperSpecialHTML))
+
+``@Html.Raw`` is a Razor HTML Helper, for more information on them see
+:doc:`html-helpers`.
+
+*******
+Models
+*******
+The ``@model`` directive is a way to indicate what model the view is using. Just
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: NTaylorMullen: Should add a section defining what a Razor directive is, much like what you do for Implicit and Explicit expressions

File: mvc/views/razor.rst [hunk 15]
```diff
@@ -1,8 +1,283 @@
...
+
+.. code-block:: c#
+
+  @await FlushAsync()
+
+**********
+@inherits
+**********
+``@inherits`` Allows you to override the class that view inherits from. By
+default the view inherits from ``System.Web.Mvc.WebViewPage``.
```
Comment (pranavkm) [inline mvc/views/razor.rst]: Views inherit from `Microsoft.AspNet.Mvc.Razor.RazorPage`

File: mvc/views/razor.rst (hunk 14 above)
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: To take it further, add a parent section `Directives` where you can define what a directive is and then list out each one and its capabilities.

File: mvc/views/razor.rst [hunk 16]
```diff
@@ -1,8 +1,283 @@
...
+
+.. code-block:: c#
+
+  @using Example.Models;
+
+************
+Code Blocks
+************
+Code blocks are a fundamental element of razor. A code block is a section where
+c# code, has been inserted directly into the html. It follows the same rules
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: no comma

File: mvc/views/razor.rst [hunk 17]
```diff
@@ -1,8 +1,283 @@
...
+that a normal c# block does. Implicit expressions do not have trailing
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: `Implicit expressions do not have trailing semicolons.` Move this to a note.

File: mvc/views/razor.rst [hunk 18]
```diff
@@ -1,8 +1,283 @@
...
+  Author: @Model.AuthorName()
+
+The ``@using`` directive is the Razor equivalent of the c# ``using`` statement.
+
+.. code-block:: c#
+
+  @using Example.Models;
+
+************
+Code Blocks
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Maybe this section should go above the Expressions section. The core of Razor is intermingling C# and Razor. By having it prior the Expressions section it leads onto easier ways to interweave code and HTML.

File: mvc/views/razor.rst [hunk 19]
```diff
@@ -1,8 +1,283 @@
...
+  }
+
+The combination of code blocks and implicit expressions allow variables to be
+created and used.
+
+.. code-block:: c#
+
+  <div id="greeting">
+    @{
+      var GuestName = "John";
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: `var guestName = "John";` Will have to refactor the `@GuestName` below as well.

File: mvc/views/razor.rst [hunk 20]
```diff
@@ -1,8 +1,283 @@
...
+    }
+
+    <h2>Greetings @GuestName</h2>
+  </div>
+
+You can mix plain text, and html inside of a code block by using
+html tags and the ``<text>`` tag. The ``<text>`` tag is a placeholder tag that
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: NTaylorMullen: Should also talk about @: since you’re talking about <text>

File: mvc/views/razor.rst [hunk 21]
```diff
@@ -1,8 +1,283 @@
...
+  @{
+    var title = "Hello!";
+    <h1>@title</h1>
+
+    var body = "some text not sitting inside of a tag";
+    <text>@body</text>
+  }
+
+*******************
+Control Structures
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Move to after Expressions

File: mvc/views/razor.rst [hunk 22]
```diff
@@ -1,8 +1,283 @@
...
+.. code-block:: c#
+
+  @*
+    Some comment
+  *@
+
+***************************
+Application Relative Paths
+***************************
+One of my favorite features of razor is the ability to use application relative
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Too personable for a reference article.

File: mvc/views/razor.rst [hunk 23]
```diff
@@ -1,8 +1,283 @@
...
+
+***************
+Tag Attributes
+***************
+Razor makes it easy to have programmatic classes and ids for you html tags.
+
+.. code-block:: c#
+
+  @{
+    var AnchorClass = "dark link";
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: `anchorClass` and `anchorId`

File: mvc/views/razor.rst [hunk 24]
```diff
@@ -1,8 +1,283 @@
...
+    var AnchorId = "main-link";
+  }
+
+  <a href="#" class="@AnchorClass" id="@AnchorId"> Link </a>
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: No need for spaces around `Link`, same below.

File: mvc/views/razor.rst [hunk 25]
```diff
@@ -1,8 +1,283 @@
...
+    bool enabled = false;
+  }
+
+  <input type="checkbox" name="agree" value="agree" checked="@agree"/>
+  <input type="checkbox" name="option" value="option" disabled="@enabled"/>
+
+.. code-block:: html
+
+  <input type="checkbox" name="agree" value="agree" checked="checked"/>
+  <input type="checkbox" name="option" value="option" disabled="disabled"/>
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Doesnt the disabled attribute get nuked just like a `null` value?

File: mvc/views/razor.rst [hunk 26]
```diff
@@ -1,8 +1,283 @@
...
+
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Should add 1 more example showcasing that you don't get the magic `true`/`false` behavior when there are multiple dynamic expressions in an attribute:

`<input custom="@true hi" />` => `<input custom="True hi" />`
`<input custom="@true hi @null" />` => `<input custom="True hi" />` Note the missing space after `hi`
`<input custom="@null @null" />` => `<input custom="" />`

File: mvc/views/razor.rst [hunk 27]
```diff
@@ -1,8 +1,283 @@
...
+independent of the controller.
+
+.. code-block:: c#
+
+  @inject IInjectedServiceInterface InjectedServiceName
+
+A more in depth guide can be found here :doc:`dependency-injection`
+
+******
+Async
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Should be collapsed into the `ImplicitExpression` section.

File: mvc/views/razor.rst [hunk 28]
```diff
@@ -1,8 +1,283 @@
...
+  <input type="checkbox" name="agree" value="agree" checked="@agree"/>
+  <input type="checkbox" name="option" value="option" disabled="@enabled"/>
+
+.. code-block:: html
+
+  <input type="checkbox" name="agree" value="agree" checked="checked"/>
+  <input type="checkbox" name="option" value="option" disabled="disabled"/>
+
+*********************
+Dependency Injection
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Should be moved to directives section.

File: mvc/views/razor.rst [hunk 29]
```diff
@@ -1,8 +1,283 @@
...
+this is so that while the main content of your page is being processed and built
+by razor, important sections containing css or your navbar can be sent out ahead
+early.
+
+.. code-block:: c#
+
+  @await FlushAsync()
+
+**********
+@inherits
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: moved to directives section.

File: mvc/views/razor.rst [hunk 30]
```diff
@@ -1,8 +1,283 @@
...
+the use of ``@await Awaitable``
+
+.. code-block:: c#
+
+  <p>@await GetNameAsync()</p>
+  <p>@await Task.FromResult("Foo")</p>
+
+In addition, inside of the layout file you can flush the page. The logic behind
+this is so that while the main content of your page is being processed and built
+by razor, important sections containing css or your navbar can be sent out ahead
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Razor

File: mvc/views/razor.rst (hunk 15 above)
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Not anymore :smile: https://github.com/aspnet/Mvc/blob/e8ab079b06c330576ad616b53be87d7e83db424e/src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs#L25

File: mvc/views/razor.rst [hunk 31]
```diff
@@ -1,8 +1,283 @@
...
+.. code-block:: c#
+
+  @await FlushAsync()
+
+**********
+@inherits
+**********
+``@inherits`` Allows you to override the class that view inherits from. By
+default the view inherits from ``System.Web.Mvc.WebViewPage``.
+
```
Comment (NTaylorMullen) [inline mvc/views/razor.rst]: Should add a bit to this section and the inject section mentioning the capabilities of `@inherits MyBaseType<TModel>`. Aka we MVC injects the current model in place of `<TModel>` for inherits and inject.

/cc @pranavkm

File: mvc/views/razor.rst (hunk 25 above)
Comment (pranavkm) [inline mvc/views/razor.rst]: It's just a regular HTML attribute. Not sure why it would be removed

Comment (NTaylorMullen) [inline mvc/views/razor.rst]: We special case `@false`

File: mvc/views/razor.rst (hunk 31 above)
Comment (Eilon) [inline mvc/views/razor.rst]: It's also not `System.Web.Mvc.WebViewPage` anymore... that's from MVC 5.

**Comment by danroth27** [comment]
@robbert229 do you want any help addressing the feedback on this PR?

**Comment by robbert229** [comment]
No thanks, I've been busy with the holidays but I've managed to resume working on it in the past few days. I should have all the feedback implemented some time tonight.

File: mvc/views/razor.rst [hunk 32]
```diff
@@ -1,8 +1,481 @@
...
+~~~~~~
+The ``@using`` directive is the Razor equivalent of the C# ``using`` statement.
+
+.. code-block:: C#
+
+  @using Example.Models
+
+
+@functions
+~~~~~~~~~~
```
Comment (robbert229) [inline mvc/views/razor.rst]: @NTaylorMullen this is new. Does it look okay?

File: mvc/views/razor.rst [hunk 33]
```diff
@@ -1,8 +1,481 @@
...
+.. code-block:: C#
+
+  @inject IInjectedServiceInterface InjectedServiceName
+
+A more in depth guide can be found here :doc:`dependency-injection`
+
+
+@await
+~~~~~~
+Razor can render the results of an asynchronous method call through
```
Comment (robbert229) [inline mvc/views/razor.rst]: I renamed this from Async to @await, it seems more appropriate. In addition is this actually a razor directive?

File: mvc/views/razor.rst [hunk 34]
```diff
@@ -1,8 +1,481 @@
...
+.. code-block:: C#
+
+  @inherits MyCustomRazorPage
+
+In addition, ``@inherits`` allows you to specify the model as well.
+
+  @inherits MyCustomRazorPage<Model>
+
+
+@section
```
Comment (robbert229) [inline mvc/views/razor.rst]: @NTaylorMullen This is also new.

File: mvc/views/razor.rst [hunk 35]
```diff
@@ -1,8 +1,481 @@
...
+
+  <input custom="@true foo"/>
+  <input custom="@null @null"/>
+
+.. code-block:: HTML
+
+  <input custom="True foo"/>
+  <input custom=""/>
+
+Layouts
```
Comment (robbert229) [inline mvc/views/razor.rst]: @NTaylorMullen This is new.

**Comment by robbert229** [comment]
@danroth @NTaylorMullen @Rick-Anderson So far I have reorganized the article as well as incorporated the feedback. Any further remarks?

**Comment by danroth27** [comment]
@robbert229 Thanks for this updated contribution. I think there are still some structural issues with this article that need to be addressed. We need to get the sections into a more logical order and add additional details. I'd like to get the engineering team to help out with this, but they are currently heads down getting ready for our next milestone release. So I think we will need to put this PR on hold until folks free up in mid Feb. I apologize for the delay.

**Comment by Rick-Anderson** [comment]
@danroth27  let me know when the team finishes and I'll do a UE pass (and possibly reorder sections).

**Comment by Rick-Anderson** [comment]
@robbert229  closing for now until I have time to work on this.

**Pull request closed without merging**